packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/pattern_item.dart

```diff
@@ -4,32 +4,50 @@
 
 import 'package:flutter/foundation.dart' show immutable;
 
+/// Enumeration of types of pattern items.
+enum PatternItemType {
+  /// A dot used in the stroke pattern for a [Polyline].
+  dot('dot'),
+
+  /// A dash used in the stroke pattern for a [Polyline].
+  dash('dash'),
+
+  /// A gap used in the stroke pattern for a [Polyline].
+  gap('gap');
+
+  const PatternItemType(this.name);
+
+  /// String name used for serialization.
+  final String name;
+}
+
 /// Item used in the stroke pattern for a Polyline.
 @immutable
 class PatternItem {
-  const PatternItem._(this._json);
+  const PatternItem._(this._type, [this._length]);
 
   /// A dot used in the stroke pattern for a [Polyline].
-  static const PatternItem dot = PatternItem._(<Object>['dot']);
+  static const PatternItem dot = PatternItem._(PatternItemType.dot);
 
   /// A dash used in the stroke pattern for a [Polyline].
   ///
   /// [length] has to be non-negative.
   static PatternItem dash(double length) {
     assert(length >= 0.0);
-    return PatternItem._(<Object>['dash', length]);
+    return PatternItem._(PatternItemType.dash, length);
   }
 
   /// A gap used in the stroke pattern for a [Polyline].
   ///
   /// [length] has to be non-negative.
   static PatternItem gap(double length) {
     assert(length >= 0.0);
-    return PatternItem._(<Object>['gap', length]);
+    return PatternItem._(PatternItemType.gap, length);
   }
 
-  final Object _json;
+  final PatternItemType _type;
+  final double? _length;
```

Comment:
It seems like we'll need subclasses that expose `length` publicly, using a similar structure to your cap changes, otherwise the platform implementation code would still need to call and interpret `toJson` to get the required info.